Warn about async-imported selector conflicts

beccadax opened the pull request:
The change in #59479 inadvertently fixed another bug in selector conflict checking: Swift did not notice when an ObjC header declared an async-imported method, but a Swift extension to the same class added another method with the same selector. Unfortunately, fixing this bug was source-breaking, and it also caused Swift to diagnose spurious conflicts between the various names that a single ObjC method might be imported with. Soften the error to a warning in Swift 5 mode and suppress the spurious diagnostics.

Fixes rdar://95887113.

beccadax committed:
Warn about async-imported selector conflicts

The change in #59479 inadvertently fixed another bug in selector conflict checking: Swift did not notice when an ObjC header declared an async-imported method, but a Swift extension to the same class added another method with the same selector. Unfortunately, fixing this bug was source-breaking, and it also caused Swift to diagnose spurious conflicts between the various names that a single ObjC method might be imported with. Soften the error to a warning in Swift 5 mode and suppress the spurious diagnostics.

Fixes rdar://95887113.

beccadax commented:
@swift-ci please smoke test

hamishknight commented on lib/Sema/TypeCheckDeclObjC.cpp:
```diff
@@ -2534,16 +2543,28 @@ bool swift::diagnoseObjCMethodConflicts(SourceFile &sf) {
+      bool breakingInSwift5 = false;
+      // Imported ObjC async methods weren't fully checked for selector
+      // conflicts until 5.7.
+      if (originalMethod->hasClangNode() && originalMethod->hasAsync())
```
Does any special consideration need to be given to the ordering to make sure `originalMethod` ends up being the async alternative instead of the imported completion handler version here?

beccadax replied:
You know, the more I look at how we select the “right” method, the less I trust it. I’m going to make this a little more robust.

hamishknight reviewed (approved):
Been a little while since I've worked in this area, I'm not sure I fully understand why this issue doesn't come up for the non-async-alternative imported version, but this LGTM modulo the question on ordering.

kavon commented on lib/Sema/TypeCheckDeclObjC.cpp:
```diff
@@ -2257,6 +2257,12 @@ namespace {
   /// Produce a deterministic ordering of the given declarations.
   struct OrderDeclarations {
     bool operator()(ValueDecl *lhs, ValueDecl *rhs) const {
+      // If one declaration is imported from ObjC and the other is native Swift,
+      // put the imported Clang one first.
+      if (lhs->hasClangNode() != rhs->hasClangNode()) {
+        return lhs->hasClangNode();
+      }
+
```
great fix!

kavon commented on lib/Sema/TypeCheckDeclObjC.cpp:
```diff
@@ -2534,16 +2543,28 @@ bool swift::diagnoseObjCMethodConflicts(SourceFile &sf) {
         if (auto accessor = dyn_cast<AccessorDecl>(originalMethod))
           originalDecl = accessor->getStorage();
 
+      // Conflicts between two imported ObjC methods are caused by the same
+      // clang decl having several Swift names.
+      if (originalDecl->hasClangNode() && conflictingDecl->hasClangNode())
+        continue;
```
I wonder this is what was needed to handle the FIXME I left behind (and I think didn't file a bug for) in `test/Inputs/clang-importer-sdk/usr/include/EffectfulProperties.h` for this name clash on import as an async getter:

```
// FIXME: this does not raise an error, and generates two identical properties!
// -(void)anotherExampleWithCompletionBlock:(void (^)(NSString *))block
//   ASYNC_NAME("getter:manifestedString()");
// -(void)finalExampleWithReplyTo:(void (^)(NSString *))block
//   ASYNC_NAME("getter:manifestedString()");
```

Could you check if this gets resolved? Otherwise I'll file a proper bug for this.

beccadax replied:
I doubt this change will affect that, because it affects how we diagnose two Swift methods with the same ObjC selector, whereas your example is two ObjC selectors with the same Swift name.

kavon reviewed (approved):
Looks good. Have a question though!

beccadax committed:
Increase determinism of selector conflict errors

Refactor ObjC conflict diagnosis code to sort conflict data more thoroughly, filter out unwanted declarations earlier, and just generally behave in ways that are more likely to work correctly.

This change increases the determinism of the ordering of diagnostics and the selection of the “correct” declaration that the others are considered to conflict with, increasing my confidence that the diagnostics will work correctly in untested corner cases or if the compiler is refactored so that declarations are recorded in a different order. It also adds a new selection rule—@objc without vs. with explicit selector—that I believe will slightly improve the diagnostics we produce. And it replaces a lot of really dodgy-looking logic that may have only worked reliably when a conflict involved exactly two methods.

beccadax commented:
@swift-ci please smoke test

beccadax commented:
@swift-ci please test source compatibility

beccadax commented:
When I looked more closely at the existing code, I became less and less convinced that it would reliably do the right thing when there were more than two methods involved in a conflict, so I’ve rewritten the function to work in a way that I think will be more reliable. This perturbed a bunch of tests (!), so I also added a new rule to the set used to choose which method “should” have the selector. It’s a bigger diff than I want, but I’m much more confident that it will work correctly in cases we don’t directly test.
